client/src/util/routes.js

```diff
@@ -0,0 +1,68 @@
+import acronym from 'pdap-design-system/images/acronym.svg';
+
+const DEFAULT_META_TAGS = new Map([
+	[
+		'description',
+		'Data and tools for answering questions about any police system in the United States',
+	],
+	['title', 'Police Data Accessibility Project'],
+	[
+		'og:description',
+		'Data and tools for answering questions about any police system in the United States',
+	],
+	['og:title', 'Police Data Accessibility Project'],
+	['og:type', 'website'],
+	['og:site_name', 'PDAP'],
+	['og:image', acronym],
+]);
+const META_PROPERTIES = [...DEFAULT_META_TAGS.keys(), 'og:url'];
+
+/**
+ * Adds meta tags by route
+ * @param {RouteLocationNormalized} to Vue router route location
+ */
+export function refreshMetaTagsByRoute(to) {
+	// Get nearest matched route that has title / meta tag overrides
+	const nearestRouteWithTitle = [...to.matched]
+		.reverse()
+		.find((route) => route?.meta?.title);
+
+	const nearestRouteWithMeta = [...to.matched]
+		.reverse()
+		.find((route) => route?.meta?.metaTags);
+
+	// Update document title
+	document.title =
+		nearestRouteWithTitle?.meta?.title ?? DEFAULT_META_TAGS.get('title');
+
+	// Update meta tags
+	Array.from(document.querySelectorAll('[data-controlled-meta]')).forEach(
+		(el) => el.parentNode.removeChild(el),
+	);
+
+	META_PROPERTIES.filter((prop) => prop !== 'title')
+		.map((prop) => {
+			const tagInRouteMetaData = nearestRouteWithMeta?.meta?.metaTags?.find(
+				(tag) => tag.property === prop,
+			);
+
+			let content;
+			switch (true) {
+				case prop === 'og:url':
+					content = `${import.meta.env.VITE_VUE_APP_BASE_URL}${to.fullPath}`;
```

Comment:
My general philosophy is that once we get to so many layers of indentation (I believe there are six here?) it may be useful to extract some of this into a separate function (or several), just to improve clarity. There's a lot going on in this code!

Reply:
> My general philosophy is that once we get to so many layers of indentation

Just curious: could you explain how indentation is an accurate measure of function complexity? I could _maybe_ see it in Python, where indentation is semantically meaningful. But in JS, indentation is just formatting.

I've refactored this to split some of it out into separate utils, but if your definition of function complexity is indentation, this has done nothing to simplify things 😅 .

Reply:
From my perspective, the changes you provided did simplify things! They're now discrete utils with their own comments, and they're organized separately so that `refreshMetaTagsByRoute` now shows more of the high level details rather than implementation. From my perspective, it's much more readable.

As far as indentation goes, in this case I used it as a short heuristic for the depth of logic expressed within a single function. The deepest level includes the function itself, the filter function in `META_PROPERTIES.filter()`, the `map` function, then the switch statement, and finally the case. So multiple levels of logic are expressed within the same scope, and it is harder to read because content is increasingly moved to the right side of the page (indeed, the deepest lines, at least on my screen, wrap around back to the left side)

To give an example of how I might do it differently, I might extract the logic for obtaining `tagInRouteMetaData` to a separate function called `getTagInRouteMetaData()` the logic for obtaining `content` to a `getContent()` function, and the logic for getting and setting the `tag` to a `getAndSetTag()` function. This keeps the containing `refreshMetaTagsByRoute(to)` cleaner, and clearly delineates the three behaviors into separate logical strands that are then integrated. 

My general coding philosophy is informed by a preference for the explicit over the concise, and a desire to separate high and low-level logic even if that involves the creation of additional functions. That's undoubtedly informed by my background in Python; from what I've seen in Javascript code, a different coding style is often preferred. But I can't easily dispose of my own biases 🙃.